### PR TITLE
feat(zip): add automatic charset detection for non-UTF8 filenames

### PR DESCRIPTION
**Pull request by hooyao**
## Summary

- Add automatic encoding detection for ZIP filenames encoded in legacy code pages (Shift-JIS, GBK, EUC-KR, Windows-1251, etc.) using UtfUnknown (Mozilla's Universal Charset Detector)
- Replace eager `FileName` string property with bytes-first `FileNameBytes` + `DecodeFileName()` method on `ZipCentralDirectoryEntry` — filenames are decoded once after detection, no trie mutation needed
- Two-level detection: per-archive fast path (statistical, high accuracy) with per-entry fallback for mixed-encoding archives
- Three-tier detection chain: UtfUnknown → system OEM code page → configurable fallback encoding
- Add `FallbackEncoding` and `EncodingConfidenceThreshold` configuration on `MountOptions`
- Add telemetry counter `zip.encoding.detections` with detection path and encoding name tags
- 16 new tests (8 unit + 8 integration) with binary-level ZIP construction for Japanese, Chinese, Korean, Cyrillic, and mixed-encoding archives

## Test plan

- [x] All 224 existing tests pass (zero regressions)
- [x] 8 new `FilenameEncodingDetectorTests` — Shift-JIS, GBK, EUC-KR, Cyrillic detection; empty/ASCII/high-threshold edge cases; per-entry detection
- [x] 8 new `EncodingIntegrationTests` — end-to-end with binary ZIPs: single-encoding (Shift-JIS, GBK), mixed-encoding (Japanese+Chinese, Korean+Cyrillic), ASCII-majority, UTF-8 flag bypass, parent directory synthesis for decoded paths
- [x] Manual test with a real Shift-JIS ZIP archive from a Japanese Windows system

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/ZipDrive.Infrastructure.Archives.Zip.Tests/ZipReaderTests.cs [hunk 1]
```diff
@@ -150,10 +150,10 @@ public async Task StreamCentralDirectoryAsync_ValidZip_YieldsAllEntries()
...
             // Assert
             entries.Should().HaveCount(4);
-            entries.Select(e => e.FileName).Should().Contain("file1.txt");
-            entries.Select(e => e.FileName).Should().Contain("file2.txt");
-            entries.Select(e => e.FileName).Should().Contain("folder/file3.txt");
-            entries.Select(e => e.FileName).Should().Contain("folder/subfolder/file4.txt");
+            entries.Select(e => e.DecodeFileName()).Should().Contain("file1.txt");
+            entries.Select(e => e.DecodeFileName()).Should().Contain("file2.txt");
+            entries.Select(e => e.DecodeFileName()).Should().Contain("folder/file3.txt");
+            entries.Select(e => e.DecodeFileName()).Should().Contain("folder/subfolder/file4.txt");
```
[the remark this answers is not in the document]
Comment (hooyao) [reply tests/ZipDrive.Infrastructure.Archives.Zip.Tests/ZipReaderTests.cs:156]: The new ArchiveStructureCache constructor parameters are optional (encodingDetector = null, fallbackEncoding = null), so existing call sites compile without changes. The .FileName references in FileReadCorrectnessTests.cs and EnduranceTest.cs are on ManifestEntry objects, not ZipCentralDirectoryEntry. All 240 tests pass.

File: src/ZipDrive.Infrastructure.Archives.Zip/FilenameEncodingDetector.cs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        for (int i = 0; i < filenameByteArrays.Count; i++)
+        {
+            filenameByteArrays[i].CopyTo(concatenated, offset);
+            offset += filenameByteArrays[i].Length;
+            if (i < filenameByteArrays.Count - 1)
+            {
+                concatenated[offset] = 0; // NUL separator
+                offset++;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (hooyao) [reply src/ZipDrive.Infrastructure.Archives.Zip/FilenameEncodingDetector.cs]: Acknowledged. ArrayPool optimization is not warranted here — detection runs once per archive on first mount, and the byte arrays are small (filenames are typically < 256 bytes each). The allocation is transient and short-lived.

File: src/ZipDrive.Infrastructure.Archives.Zip/ZipTelemetry.cs [hunk 3]
```diff
@@ -26,4 +26,8 @@ internal static class ZipTelemetry
     internal static readonly Counter<long> BytesExtracted =
         Meter.CreateCounter<long>("zip.bytes_extracted", unit: "By",
             description: "Total bytes extracted from ZIP archives");
+
+    public static readonly Counter<long> EncodingDetections =
+        Meter.CreateCounter<long>("zip.encoding.detections",
+            description: "Encoding detection outcomes for ZIP filenames");
```
[the remark this answers is not in the document]
Comment (hooyao) [reply src/ZipDrive.Infrastructure.Archives.Zip/ZipTelemetry.cs]: Fixed in a6b4b06. Reverted ZipTelemetry and EncodingDetections back to internal. Added InternalsVisibleTo for ZipDrive.Infrastructure.Caching and the test project instead. API surface is now consistent — all ZipTelemetry members are internal.

File: src/ZipDrive.Infrastructure.Archives.Zip/Formats/ZipCentralDirectoryEntry.cs [hunk 4]
```diff
@@ -172,12 +175,27 @@ public readonly record struct ZipCentralDirectoryEntry
...
+    /// <param name="encoding">
+    /// Encoding to use. If null, defaults to UTF-8 when the UTF-8 flag (bit 11) is set,
+    /// otherwise Code Page 437 (DOS).
+    /// </param>
+    /// <returns>The decoded filename string.</returns>
+    public string DecodeFileName(Encoding? encoding = null)
+    {
+        encoding ??= IsUtf8 ? Encoding.UTF8 : Encoding.GetEncoding(437);
+        return encoding.GetString(FileNameBytes);
+    }
```
[the remark this answers is not in the document]
Comment (hooyao) [reply src/ZipDrive.Infrastructure.Archives.Zip/Formats/ZipCentralDirectoryEntry.cs:199]: Fixed in a6b4b06. CP437 is now cached as a private static readonly field to avoid repeated Encoding.GetEncoding(437) dictionary lookups.

File: src/ZipDrive.Infrastructure.Archives.Zip/FilenameEncodingDetector.cs [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                return null;
+
+            Encoding oemEncoding = Encoding.GetEncoding(oemCodePage);
+            string decoded = oemEncoding.GetString(bytes);
+
+            // Reject if decoding produces replacement characters
+            if (decoded.Contains('\uFFFD'))
+            {
+                _logger?.LogDebug(
+                    "System OEM code page {CodePage} produced replacement characters, rejecting",
```
[the remark this answers is not in the document]
Comment (hooyao) [reply src/ZipDrive.Infrastructure.Archives.Zip/FilenameEncodingDetector.cs]: Fixed in a6b4b06. Now uses DecoderFallback.ExceptionFallback with a try/catch for DecoderFallbackException instead of scanning the decoded string for U+FFFD. More correct and avoids unnecessary string allocation when the encoding is invalid.

File: src/ZipDrive.Infrastructure.Caching/ArchiveStructureCache.cs [hunk 6]
```diff
@@ -189,6 +207,109 @@ private async Task<CacheFactoryResult<ArchiveStructure>> BuildStructureAsync(
...
+            return;
+        }
+
+        // Phase 2: Per-archive detection
+        List<byte[]> allBytes = buffer.Select(e => e.CdEntry.FileNameBytes).ToList();
+        DetectionResult? archiveResult = _encodingDetector.DetectEncoding(allBytes);
+
+        if (archiveResult != null)
+        {
+            _logger.LogInformation(
```
[the remark this answers is not in the document]
Comment (hooyao) [reply src/ZipDrive.Infrastructure.Caching/ArchiveStructureCache.cs]: Fixed in a6b4b06. Changed archive-level detection success log from LogInformation to LogDebug. Per-entry fallback mode activation remains at LogInformation since it is more noteworthy.

File: tests/ZipDrive.Infrastructure.Archives.Zip.Tests/EncodingIntegrationTests.cs [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+            ("데이터/설정파일.cfg", 949),
+            ("Документы/отчёт.txt", 1251),
+            ("Настройки/конфигурация.cfg", 1251),
+        });
+
+        // Act
+        ArchiveStructure structure = await BuildStructureAsync(zipPath);
+
+        // Assert — should parse without errors
+        structure.EntryCount.Should().BeGreaterThanOrEqualTo(5);
```
[the remark this answers is not in the document]
Comment (hooyao) [reply tests/ZipDrive.Infrastructure.Archives.Zip.Tests/EncodingIntegrationTests.cs:317]: Fixed in a6b4b06. Mixed-encoding tests now verify that decoded trie keys contain CJK, Korean, or Cyrillic Unicode characters (not CP437 mojibake). Uses Unicode range checks to confirm the encoding detection actually worked.

File: openspec/changes/zip-filename-encoding-detection/tasks.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+- [x] 8.3 In `Program.cs`: update `ArchiveStructureCache` registration to pass `IFilenameEncodingDetector` and resolved `Encoding`
+
+## 9. Fix Existing Test Compilation
+
+- [x] 9.1 Update `tests/ZipDrive.TestHelpers/VfsTestFixture.cs`: pass `IFilenameEncodingDetector` and `Encoding.UTF8` to `ArchiveStructureCache` constructor
+- [x] 9.2 Update `tests/ZipDrive.Domain.Tests/ZipVirtualFileSystemTests.cs`: update `ArchiveStructureCache` construction
+- [x] 9.3 Update `tests/ZipDrive.EnduranceTests/EnduranceTest.cs`: update `ArchiveStructureCache` construction
+- [x] 9.4 Update all `ZipReaderTests.cs` assertions: migrate `.FileName` to `.DecodeFileName()`
+- [x] 9.5 Update `FileReadCorrectnessTests.cs` if it references `ZipCentralDirectoryEntry.FileName`
+- [x] 9.6 Build and run all existing tests — verify 196+ tests pass (zero regressions)
```
[the remark this answers is not in the document]
Comment (hooyao) [reply openspec/changes/zip-filename-encoding-detection/tasks.md]: Same as comment 1: the new ArchiveStructureCache constructor parameters are optional (encodingDetector = null, fallbackEncoding = null), so existing test call sites compile without changes. The .FileName references in the listed test files are on ManifestEntry or other types, not ZipCentralDirectoryEntry. Tasks are correctly marked complete — all 240 tests pass.
